Extended tern plugin

alperozisik opened the pull request:
Exposed some functionality in tern_worker via c9-plugin tern.
- Exposed adding & removing plugins.
- Exposed getting list of definitions loaded by tern_worker (instead of
  tern c9-plugin as a separate functionality)
- Improved adding multiple def performance (tern server resets once)
- Exposed changing tern server options
- Exposed changing tern request options
- It is possible to define default tern-plugins and defs via client-default. Current implementation does not require to client-default to be updated.

alperozisik committed:
Extended tern plugin

Exposed some functionality in tern_worker via c9-plugin tern.
It is now possible to:
define default tern-plugins and defs.
* Exposed adding & removing plugins.
* Exposed getting list of definitions loaded by tern_worker (instead of
tern c9-plugin as a separate functionality)
* Improved adding multiple def performance (tern server resets once)
* Exposed changing tern server options
* Exposed changing tern request options

lennartcl commented on tern.js:
```diff
@@ -96,6 +203,48 @@ define(function(require, exports, module) {
+             * @property {String} name - name of the plugin
+             * @property {boolean} enabled - Setting it false marks plugin for removal
+             * @property {String} path - Parameter to provide while loading a new plugin
+             */
```
Comment without member?

alperozisik replied:
Taken it from [@typedef](http://usejsdoc.org/tags-typedef.html) example.
As you can see in the example it is possible to create some documentation comments without member.
My Mistake forgotten to update line 223

lennartcl replied:
Ah, it's a typedef. Missed that. Please move it out of the members section then, above the freezePublicApi() call.

lennartcl commented on worker/tern_worker.js:
```diff
@@ -124,6 +132,67 @@ handler.init = function(callback) {
+               pluginExecResult = require(targetPluginInfo.path);
+               pluginToWorkWith = ternWorker.options.plugins[targetPluginInfo.name] = TERN_PLUGINS[targetPluginInfo.name] =  pluginExecResult && targetPluginInfo.enabled;
+               
+               //add special condition for architectResolver
```
:+1:

lennartcl commented on worker/tern_worker.js:
```diff
@@ -177,7 +246,7 @@ handler.analyze = function(value, ast, callback, minimalAnalysis) {
     addTernFile(this.path, value);
 
-    architectResolver.onReady(function() {
+    architectResolver && architectResolver.onReady(function() {
```
What happens if this is not set yet at init time?

alperozisik replied:
It is possible to not use architectResolver tern-plugin. The code there within the ternWorker event has reference to the object.
It is more safe to guard it

lennartcl replied:
But if it's only initialized with https://github.com/c9/c9.ide.language.javascript.tern/pull/1#discussion-diff-36980234R178, will `onReady()` always be called?

alperozisik replied:
I do not know much about architectResolver tern-plugin.
This code block resides in _handler.analyze_ method, which is an event to be called for handling the code changes to be analyzed. As the existing code calls this every time, so the change there should call it every time as the _architectResolver_ object is present

lennartcl replied:
Ok, in that context, it's good actually, thanks for pointing that out.

lennartcl commented on tern.js:
```diff
@@ -38,11 +38,72 @@ define(function(require, exports, module) {
+            
+            
+            var defsToAdd =[], defIndex, d;
+            for(defIndex in ternOptions.defs) {
```
- Sorry to bug you about this, but please check your PR against our [coding standards](https://github.com/c9/core/blob/master/docs/CODING_STANDARDS.md), particularly looking for:
- space after `=` and keywords like `for` and `if`
- declarations on separate lines (this is perhaps a soft rule, but it's especially good to do this when initializing in declarations)
- use double quotes `"` for string literals

lennartcl commented:
@alperozisik Please check the code conventions comment above and I think this should be good to go. I like your approach, extending the public API of the UI-side tern plugin rather than just tern_worker.

alperozisik commented:
OK,
I have read briefly the code convention rules of C9.
It seams that the things you mentioned is most likely could be done by formatting whole document through _JS Beautify_.
If you are OK with that, I will do that and update?

lennartcl commented:
@alperozisik Beautification on selections may work, but please don't format the whole document, that would likely create many whitespace and unrelated changes. In general we try to adhere to these conventions manually, and sometimes that's painful, but beautifiers don't seem to provide a full solution :(

lennartcl commented:
@alperozisik Let me know what you think. I would really appreciate if you can at least do the little things of https://github.com/c9/c9.ide.language.javascript.tern/pull/1#discussion_r36980880. Other than that your PR looks good. I'd like to test it with our code base and then merge it.

alperozisik commented:
Ok, I will update it till tomorrow evening

alperozisik committed:
Updated structure to match coding standarts

In addition fixed some comment

alperozisik commented:
@lennartcl updated PR with new commit

alperozisik committed:
Fixed missing implementation for Doc URL in results

in def files URL’s for documentations are recorded with !url property
and this information is passed with url property.
Passing that url was missing in implementation
Also editor has cleared just whitespace empty lines

lennartcl commented:
Great, thanks! This has been merged in upstream and released. It will soon land here.
